### PR TITLE
Sta/geopricing

### PR DESCRIPTION
**Pull request by SamTaz**
Hello, part of it is still WIP but since the notebook is pretty long I am making a PR.

You can review the first 3 parts (out of 4) about:
- Computing nr of competitors for each store
- Computing the price index of each store
- Clustering stores based on the previously computed features
- Last part using the clusters is still WIP

### CODE REVIEW

File: retail/geopricing/Geopricing.ipynb [hunk 1]
```diff
@@ -0,0 +1,2214 @@
...
+    "    scope=tt.scope.origin(lvl[\"ProductId\"], lvl[\"StoreId\"]),\n",
+    ")\n",
+    "m[\"Price Index\"] = price_index_numerator / m[\"Margin.SUM\"]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can now have the price index per store and product, let's visualize it per store."
```
Comment (HuifangYeo) [inline retail/geopricing/Geopricing.ipynb:1260]: "Now that we have..., let's visualize it..."
or
".... and product. Let's visualize..."

File: retail/geopricing/Geopricing.ipynb [hunk 2]
```diff
@@ -0,0 +1,2214 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 32,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<div>\n",
+       "<style scoped>\n"
```
Comment (HuifangYeo) [inline retail/geopricing/Geopricing.ipynb:1613]: actually outputting the previous query already gives the dataframe. Why do we need to output to CSV and to read it again? Can't we just refer to the stores_competitors_count_per_area dataframe instead?

Comment (HuifangYeo) [inline retail/geopricing/Geopricing.ipynb:1699]: can we use price_index_per_store dataframe instead of writing to csv and reading it again?

Comment (HuifangYeo) [inline retail/geopricing/Geopricing.ipynb:1843]: Just wondering if there is any chart in atoti that can fulfill the same objective as the plot used here...

File: retail/geopricing/Geopricing.ipynb [hunk 3]
```diff
@@ -0,0 +1,2214 @@
...
+    "        lvl[\"StoreId\"], lvl[\"CompetitorStoreId\"]\n",
+    "    ),\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will then use an atoti query to quickly calculate a pandas dataframe containing columns with the number of competitors in each area, and stores as rows."
```
[the remark this answers is not in the document]
Comment (SamTaz) [reply retail/geopricing/Geopricing.ipynb:669]: I don't exactly know how to put it otherwise, the output of queries is always a pandas dataframe, we cannot chose otherwise at the moment.

[the remark this answers is not in the document]
Comment (HuifangYeo) [reply retail/geopricing/Geopricing.ipynb:669]: but to say calculate a pandas dataframe sounds odd... do we calculate dataframe?

File: retail/geopricing/Geopricing.ipynb [hunk 4]
```diff
@@ -0,0 +1,2214 @@
...
+    "margin_per_product_and_store_df = pd.read_csv(\n",
+    "    \"https://data.atoti.io/notebooks/geopricing/margin_per_product_store.csv\", sep=\";\"\n",
+    ")\n",
+    "margin_per_product_and_store = session.read_pandas(\n",
+    "    margin_per_product_and_store_df,\n",
+    "    keys=[\"ProductId\", \"StoreId\"],\n",
+    "    store_name=\"MarginPerProductAndStore\",\n",
+    ")"
+   ]
+  },
```
[the remark this answers is not in the document]
Comment (SamTaz) [reply retail/geopricing/Geopricing.ipynb:932]: Done, it's indeed simpler now.
Though note that the joining stores cell took less than one second to execute before, it now takes about 30...

File: retail/geopricing/Geopricing.ipynb (hunk 2 above)
Comment (SamTaz) [reply retail/geopricing/Geopricing.ipynb]: Hm you're right, it was done like that because this was performed in different notebooks before.

**Comment by SamTaz** [comment]
> Just curious, is this notebook demonstrating how atoti can work with other Python libraries?
> It feels like atoti is only used to derive interim data here.

Hello,
Indeed, one of the requirements for this notebook was to have something using both and machine learning libraries.
In the end you have more atoti in the simulations part now

File: retail/geopricing/Geopricing.ipynb [hunk 5]
```diff
@@ -0,0 +1,2064 @@
...
+   "outputs": [],
+   "source": [
+    "price_index_per_store.to_csv(\"FeaturesPriceIndexPerStore.csv\", index=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Stores clustering using price index and competitors number features"
```
Comment (HuifangYeo) [inline retail/geopricing/Geopricing.ipynb:1457]: Can we have more explanation on this sections?
How should we interpret the charts?

File: retail/geopricing/Geopricing.ipynb [hunk 6]
```diff
@@ -0,0 +1,2064 @@
...
+    "    right_on=\"StoreId\",\n",
+    "    how=\"left\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "jupyter": {
+     "sourc
```
Comment (HuifangYeo) [inline retail/geopricing/Geopricing.ipynb:2032]: What's the objective of this simulation? What kind of decision can we make based on this?

Comment (fabiencelier) [inline retail/geopricing/Geopricing.ipynb:1892]: Can you explain the pricing strategy and what's the link to the clusters ?
I guess we will apply the same strategy per cluster and align on concurrent prices in clusters with more close concurrents.

Comment (fabiencelier) [inline retail/geopricing/Geopricing.ipynb:1920]: This external function `optimize_prices` is very opaque, and it is followed by a simulation which is also difficult to understand.
What you could do is add well chosen example such as "This product in this store used be this price but because the store is in this competitive cluster we have changed the price to this price closer to the concurrent prices"

File: retail/geopricing/Geopricing.ipynb [hunk 7]
```diff
@@ -0,0 +1,2064 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Geopricing with atoti\n",
+    "In this notebook we will explore another pricing use case combining machine learning algorithms and atoti.  \n",
+    "We will start by creating clusters of stores depending on their local competition and current prices, then simulate various pricing strategies with atoti.\n",
+    "\n",
+    "### Dependencies\n",
+    "A few libraries will need to be installed for this notebook to work:  \n",
+    "- conda install folium  \n",
```
Comment (fabiencelier) [inline retail/geopricing/Geopricing.ipynb:13]: you can oneline it to `conda install folium scipy scikit-learn matplotlib seaborn`, but I don't know if it is a best practice or not

File: retail/geopricing/Geopricing.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (fabiencelier) [reply retail/geopricing/Geopricing.ipynb:1883]: Also can you reinject these cluster into the cube and do some viz to illustrate ?

I don't know how to join them exactly but something like this maybe:

```python
clusters_df = stores_features[["cluster"]].reset_index()
clusters_df.StoreId = clusters_df.StoreId.astype("int32")
cluster_store = session.read_pandas(
    clusters_df,
    keys=["StoreId"],
    store_name = "Clusters"
)
stores_store.join(cluster_store)
```

price index

Comment (fabiencelier) [reply retail/geopricing/Geopricing.ipynb:2032]: You could plot something showing how the clustering have impacted the strategy. Can't find anything relevant but something like this:

selling price

File: retail/geopricing/Geopricing.ipynb (hunk 7 above)
Comment (tibdex) [reply retail/geopricing/Geopricing.ipynb:13]: One liners are a best practice as they allow Conda to check compatible versions of transitive dependencies only once.

File: retail/geopricing/Geopricing.ipynb (hunk 6 above)
Comment (SamTaz) [reply retail/geopricing/Geopricing.ipynb]: The strategy is very simple, decrease price if there is a lot of competition, increase it otherwise.
I've added a line mentioning that.

**Comment by SamTaz** [comment]
Hello, I processed all your comments ! You can still contact me on skype if you need help on this particular notebook.

File: retail/geopricing/Geopricing.ipynb [hunk 8]
```diff
@@ -0,0 +1,2214 @@
...
+    "    \"https://data.atoti.io/notebooks/geopricing/competitor_stores.csv\", sep=\";\"\n",
+    ")\n",
+    "competitor_stores_df.head(5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can already visualize on a map our stores (in red) and those of the competition (in blue)."
```
[the remark this answers is not in the document]
Comment (SamTaz) [reply retail/geopricing/Geopricing.ipynb:301]: France is beautiful right ?

[the remark this answers is not in the document]
Comment (fabiencelier) [reply retail/geopricing/Geopricing.ipynb:301]: > I love this map! If only there's a way we can publish this as a widget to include in our dashboard as well... @rcolle / @tibdex / @fabiencelier ?

We should probably wait for ActiveUI to put back the maps widget, there used to be one.

> France is beautiful right ?

File: retail/geopricing/main.ipynb [hunk 9]
```diff
@@ -0,0 +1,2480 @@
...
+      "\n",
+      "You can hide this message by setting the ATOTI_HIDE_EULA_MESSAGE environment variable to True.\n"
+     ]
+    }
+   ],
+   "source": [
+    "import atoti as tt\n",
+    "import pandas as pd\n",
+    "\n",
+    "from utils import utils"
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:39]: Not a big fan of this. We then have no clues about what your imported methods do.

File: retail/geopricing/main.ipynb [hunk 10]
```diff
@@ -0,0 +1,2480 @@
...
...luZm8tc2lnbiIsICJpY29uQ29sb3IiOiAid2hpdGUiLCAibWFya2VyQ29sb3IiOiAicmVkIiwgInByZWZpeCI6ICJnbHlwaGljb24ifQogICAgICAgICAgICApOwogICAgICAgICAgICBtYXJrZXJfZmJhOGI2MjFjYmEwNGJlMWEwMGIxNTU1M2FhNzY2MTkuc2V0SWNvbihpY29uXzlkNmU4Mjk5NjE0YjQ5NGViMTVkNzQwNTZkZTg0YjllKTsKICAgICAgICAKICAgIAogICAgICAgIHZhciBwb3B1cF9hZGQ1ZGVmYzRlMjg0Yjk3OTYyNjkwYWYyZjE2N2M5MCA9IEwucG9wdXAoeyJtYXhXaWR0aCI6ICIxMDAlIn0pOwoKICAgICAgICAKICAgICAgICAgICAgdmFyIGh0bWxfMDQ3MzM2NTE1NTAxNDkxMGJjZDc4ZTcyNTE2MzE3OTUgPSAkKGA8ZGl2IGlkPSJodG1sXzA0NzMzNjUxNTUwMTQ5MTBiY2Q3OGU3MjUxNjMxNzk1IiBzdHlsZT0id2lkdGg6IDEwMC4wJTsgaGVpZ2h0OiAxMDAuMCU7Ij5BY3RpdmVWaWFtQEJhc3RpYVs0Mi42OTczOzkuNDUwOV08L2Rpdj5gKVswXTsKICAgICAgICAgICAgcG9wdXBfYWRkNWRlZmM0ZTI4NGI5Nzk2MjY5MGFmMmYxNjdjOTAuc2V0Q29udGVudChodG1sXzA0NzMzNjUxNTUwMTQ5MTBiY2Q3OGU3MjUxNjMxNzk1KTsKICAgICAgICAKCiAgICAgICAgbWFya2VyX2ZiYThiNjIxY2JhMDRiZTFhMDBiMTU1NTNhYTc2NjE5LmJpbmRQb3B1cChwb3B1cF9hZGQ1ZGVmYzRlMjg0Yjk3OTYyNjkwYWYyZjE2N2M5MCkKICAgICAgICA7CgogICAgICAgIAogICAgCjwvc2NyaXB0Pg== onload=\"this.contentDocument.open();this.contentDocument.write(atob(this.getAttribute('data-html')));this.contentDocument.close();\" allowfullscreen webkitallowfullscreen mozallowfullscreen></iframe></div></div>"
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:341]: Why is this?

File: retail/geopricing/main.ipynb [hunk 11]
```diff
@@ -0,0 +1,2480 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To assess whether a store has strong competition or not, we will count how many competitor stores there are in an area of 1kms, 5kms, 10kms and 20kms.  \n",
+    "In order to to that, we first need to compute distances between stores then count the number of competitor stores for each of the kilometers values above.\n",
+    "\n",
+    "### Computing distances between stores\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:364]: Why do you need atoti to do this? It seems like this could all be done easily using pandas.

File: retail/geopricing/main.ipynb [hunk 12]
```diff
@@ -0,0 +1,2480 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "competitorStore_store = session.read_csv(\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:836]: The variable name sounds ridiculous ...
Do you need to suffix all store variables with `store`?

File: retail/geopricing/main.ipynb [hunk 13]
```diff
@@ -0,0 +1,2480 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "price_index_cube = session.create_cube(competitorPrices_store, \"CompetitorPricesCube\")"
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:930]: There were no explanations about what happened until now (store loading, joins, cube creation)

File: retail/geopricing/main.ipynb [hunk 14]
```diff
@@ -0,0 +1,2480 @@
...
+    "Let's create some measures using _Margin_ and _SellingPrice_ from the referenced stores."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "m[\"Margin.SUM\"] = tt.agg.sum(margin_per_product_and_store[\"Margin\"])"
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1335]: Isn't this wrong?
Since CompetitorsPrices is your base, you will multiply your margin by the number of competitors.
For instance

```
CompetitorsPrices
Product1,Competitor1,MyStore,1
Product1,Competitor2,MyStore,2
```

```
MarginPerProductAndStore
Product1,MyStore,10
```

`Margin.SUM` for Product1 will be equal to 20. Is that really what you want?

File: retail/geopricing/main.ipynb [hunk 15]
```diff
@@ -0,0 +1,2480 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "m[\"Product Price\"] = tt.agg.mean(sellingPrices_store[\"SellingPrice\"])"
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1344]: You can use single value (https://docs.atoti.io/0.4.1/lib/atoti.html?highlight=single_value#atoti.agg.single_value).
Actually you can even directly use `sellingPrices_store["SellingPrice"]` in the following measure

**Comment by rcolle** [comment]
Untitled

File: retail/geopricing/main.ipynb (hunk 9 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:39]: Does it help if I do the below change:
- rename utils.py to geo_utils.py
- import the geo_utils before its usage?

File: retail/geopricing/main.ipynb (hunk 11 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:364]: I thought the same but I guess it was to somehow to make use of atoti in this notebook

File: retail/geopricing/main.ipynb (hunk 12 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:836]: I removed the suffix.

File: retail/geopricing/main.ipynb [hunk 16]
```diff
@@ -0,0 +1,3792 @@
...
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "distance_matrix = session.read_pandas(\n",
+    "    stores_distances_matrix,\n",
+    "    keys=[\"StoreId\", \"CompetitorStoreId\"],\n",
+    "    store_name=\"DistanceMatrix\",\n",
+    "    types={\"StoreId\": tt.types.INT, \"CompetitorStoreId\": tt.types.INT,},\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:610]: The fields are displayed as floating point numbers even though they are declared as `int`s.
1. Is there a bug or is it a leftover from a previous run?
2. Why did you need to specify the type? What's the pandas type?

File: retail/geopricing/main.ipynb [hunk 17]
```diff
@@ -0,0 +1,3792 @@
...
+    "It’s a good practice to do this kind of fact-level preprocessing outside atoti, as it avoids redoing these computations on every row over and over, slowing down every query."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#  We used pandas to read the selling price here as we will be using it again for price optimisation in the later section.\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:638]: Aren't you doing this to pre-process the data?

File: retail/geopricing/main.ipynb [hunk 18]
```diff
@@ -0,0 +1,3792 @@
...
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "products = session.read_csv(\n",
+    "    \"s3://data.atoti.io/notebooks/geopricing/products_info.csv\",\n",
+    "    sep=\";\",\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1035]: Isn't the separator automatically discovered by atoti?

File: retail/geopricing/main.ipynb [hunk 19]
```diff
@@ -0,0 +1,3792 @@
...
+     },
+     "execution_count": 15,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "stores = session.read_pandas(\n",
+    "    stores_df,\n",
+    "    #     sep=\";\",\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1185]: leftover

File: retail/geopricing/main.ipynb [hunk 20]
```diff
@@ -0,0 +1,3792 @@
...
+     },
+     "execution_count": 16,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "competitorStore = session.read_pandas(\n",
+    "    competitor_stores_df,\n",
+    "    #     sep=\";\",\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1343]: leftover

File: retail/geopricing/main.ipynb [hunk 21]
```diff
@@ -0,0 +1,3792 @@
...
+    "productSales.join(\n",
+    "    competitorPrices, mapping={\"ProductId\": \"ProductId\", \"StoreId\": \"StoreId\"}\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To have more efficient joins, we join the _Products_ and _Stores_ to the base store since this is the most direct reference. "
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1376]: What do you mean by `efficient` joins?

File: retail/geopricing/main.ipynb [hunk 22]
```diff
@@ -0,0 +1,3792 @@
...
+   "source": [
+    "productSales.join(products, mapping={\"ProductId\": \"ProductId\"})\n",
+    "productSales.join(stores, mapping={\"StoreId\": \"StoreId\"})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As _CompetitorPrices_ store is more granular than the _CompetitorStore_, we join the latter to _CompetitorPrices_."
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1393]: Is this the right way to put it? Aren't you "simply" enriching the CompetitorPrices store with details about the competitors stores?
BTW, can't you call retail stores "shops" to distinguish them from database stores?

File: retail/geopricing/main.ipynb [hunk 23]
```diff
@@ -0,0 +1,3792 @@
...
+    "competitorPrices.join(\n",
+    "    competitorStore, mapping={\"CompetitorStoreId\": \"CompetitorStoreId\"}\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "_DistanceMatrix_ contains both the _StoreId_ and _CompetitorStoreId_.  \\\n",
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1411]: Too many `\`

File: retail/geopricing/main.ipynb [hunk 24]
```diff
@@ -0,0 +1,3792 @@
...
+    "competitorStore.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "_ProductSales_ will be our base store as we want to be able to see all the products available in our stores.  \n",
+    "Competitors may or may not have the same products as our product catalogue. We are only interested in the products that we are also selling.  \n",
+    "Hence we will join the _CompetitorPrices_ store to our _ProductSales_ store by _ProductId_ and _StoreId_."
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1358]: Do you want to mention the fact that it's a many to many relationship, and not a simple "data enrichment" like the other joins? I dont know if this is worth putting forward or not.

File: retail/geopricing/main.ipynb [hunk 25]
```diff
@@ -0,0 +1,3792 @@
...
+       "<text text-anchor=\"start\" x=\"1017\" y=\"-489.8\" font-family=\"Times New Roman,serif\" font-style=\"italic\" font-size=\"14.00\">int</text>\n",
+       "<text text-anchor=\"start\" x=\"1033\" y=\"-489.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\"> </text>\n",
+       "<polygon fill=\"none\" stroke=\"black\" points=\"863,-437 863,-474 1083,-474 1083,-437 863,-437\"/>\n",
+       "<text text-anchor=\"start\" x=\"874\" y=\"-452.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\">CompetitorStoreName </text>\n",
+       "<text text-anchor=\"start\" x=\"1004\" y=\"-452.8\" font-family=\"Times New Roman,serif\" font-style=\"italic\" font-size=\"14.00\">String</text>\n",
+       "<text text-anchor=\"start\" x=\"1040\" y=\"-452.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\"> </text>\n",
+       "<polygon fill=\"none\" stroke=\"black\" points=\"863,-400 863,-437 1083,-437 1083,-400 863,-400\"/>\n",
+       "<text text-anchor=\"start\" x=\"874\" y=\"-415.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\">CompetitorStoreAdress </text>\n",
+       "<text text-anchor=\"start\" x=\"1012\" y=\"-415.8\" font-
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1875]: Too many `\`

File: retail/geopricing/main.ipynb [hunk 26]
```diff
@@ -0,0 +1,3792 @@
...
+    "Normally, we can have the selling price, cost price and quantity of the product sold. Margin hence can be derived, and we can consequently see the impact on margin when price index changes.  \n",
+    "In this notebook however, we simplify the use case by having only the SellingPrice and the total Margin for each product by stores.  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We multiply SellingPrice by Margin in the dataframe before loading it into atoti. This value will be used to compute the price index of the stores in the later section.  \n",
+    "It’s a good practice to do this kind of fact-level preprocessing outside atoti, as it avoids redoing these computations on every row over and over, slowing down every query."
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:629]: Do you really need to do that? Is this only for the following computation?
```
m["SellingPrice_x_Margin.SUM"] / m["CompetitorPrice.MEAN"]
```

If so there's really no point in doing so. You're already doing a division at this level, you might as well do an extra multiplication, it's basically free.

File: retail/geopricing/main.ipynb (hunk 25 above)
Comment (Skyscimitar) [inline retail/geopricing/main.ipynb:2099]:
```
m["Threshold"] = 1
```
This is actually sufficient, I went a bit overboard with the `single_value` in the suggestion I sent you.

Comment (Skyscimitar) [inline retail/geopricing/main.ipynb:2139]:
```
m["Distance<Thesh"] = tt.agg.single_value(
  tt.where(m["Competitor distance KM.VALUE"] < m["Threshold"], m["Nr Competitors"], 0),
  scope=tt.scope.origin(lvl["ProductId"]),
)
```
Same as above

Comment (Skyscimitar) [inline retail/geopricing/main.ipynb:2086]: No longer necessary, or at least needs re-adapting to the measures which still need the `single_value` aggregation function.

File: retail/geopricing/main.ipynb [hunk 27]
```diff
@@ -0,0 +1,3792 @@
...
+    "    distance_matrix,\n",
+    "    mapping={\"CompetitorStoreId\": \"CompetitorStoreId\", \"StoreId\": \"StoreId\"},\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's see the final design of our cube."
```
Comment (rcolle) [inline retail/geopricing/main.ipynb:1432]: I find that all the loading / joins is now too verbose :-) Sorry for being picky.
Maybe you can explain that that you pick the sales as the base store because these are your facts, and then say that you're enriching them with extra information to create a star / snowflake schema, that can then be visualized with `cube.schema` (which I find quite clear).

As mentioned earlier you may probably need to talk a bit about the many to many relationship between sales and competitors as this is the only non-standard enrichment.

File: retail/geopricing/main.ipynb (hunk 25 above)
Comment (rcolle) [inline retail/geopricing/main.ipynb:2100]: That's horrible :-)
If these are workarounds for core issues, we need to fix them and release a quick bug fix so that your notebook doesn't look bad. Can you file a ticket with the issues encountered there?

Comment (Skyscimitar) [inline retail/geopricing/main.ipynb:2095]: `m["Threshold"] =1` is sufficient here.

Comment (Skyscimitar) [inline retail/geopricing/main.ipynb:2136]: Same as above, you can use the `m["Competitor distance KM.VALUE"]` measure here, I went a bit overboard with the use of the `single_value` function in the suggestions I sent you.

Comment (Skyscimitar) [inline retail/geopricing/main.ipynb:2085]: This will need adapting to the measures which still require the use of the `single_value` aggregation function.

Comment (rcolle) [reply retail/geopricing/main.ipynb]: I think I messed up the comment ...
This was supposed to refer to the cell where you define `m["1"] = 1` and also a single value agg on top of a `.VALUE` measure

File: retail/geopricing/main.ipynb (hunk 16 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:610]: It should be a bug - https://github.com/atoti/atoti/issues/86. 
There was inconsistency in the data type when read by the session. 
Hence it is safer to specify the type.

image

By specifying the type, even though the displayed value is a floating point number, the displayed type is still `int`.
image

By casting the pandas type to int32, we can see that the type of the field is now int without having to specify the type.
However the displayed table still shows floating point number.
image

File: retail/geopricing/main.ipynb (hunk 17 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:638]: It's meant for price optimization using source simulation.

File: retail/geopricing/main.ipynb (hunk 21 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:1376]: Initially I thought I should join the Product store to the CompetitorPrices store and ProductSales store.
However, it is sufficient to join it directly to the ProductSales. I tried to give some explanation but I guess it's not really well explained. I'll remove the explanations since the notebook is already very long.

File: retail/geopricing/main.ipynb (hunk 22 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:1393]: I'll remove the descriptions on the join and rename the stores to outlets.

File: retail/geopricing/main.ipynb (hunk 23 above)
Comment (HuifangYeo) [reply retail/geopricing/main.ipynb:1411]: removed line.

File: retail/geopricing/main.ipynb [hunk 28]
```diff
@@ -0,0 +1,3367 @@
...
+    "# config = tt.config.create_config(port=9999)\n",
+    "session = tt.create_session()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We load the distance matrix into a data store. This will facilitate the computation of the number of competitors per area radius.  \n",
+    "Though not implemented in this notebook, we can also use it for analysis of price index against the competitors within a distance range."
```
Comment (xpilasav) [inline retail/geopricing/main.ipynb:523]: Remove this line? not useful as not done...

File: retail/geopricing/main.ipynb [hunk 29]
```diff
@@ -0,0 +1,3367 @@
...
+       "<text text-anchor=\"start\" x=\"824\" y=\"-304.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\">CompetitorOutletCountry </text>\n",
+       "<text text-anchor=\"start\" x=\"972\" y=\"-304.8\" font-family=\"Times New Roman,serif\" font-style=\"italic\" font-size=\"14.00\">String</text>\n",
+       "<text text-anchor=\"start\" x=\"1008\" y=\"-304.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\"> </text>\n",
+       "<polygon fill=\"none\" stroke=\"black\" points=\"813,-252 813,-289 1037,-289 1037,-252 813,-252\"/>\n",
+       "<text text-anchor=\"start\" x=\"824\" y=\"-267.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\">CompetitorOutletLatitude </text>\n",
+       "<text text-anchor=\"start\" x=\"971\" y=\"-267.8\" font-family=\"Times New Roman,serif\" font-style=\"italic\" font-size=\"14.00\">double</text>\n",
+       "<text text-anchor=\"start\" x=\"1011\" y=\"-267.8\" font-family=\"Times New Roman,serif\" font-size=\"14.00\"> </text>\n",
+       "<polygon fill=\"none\" stroke=\"black\" points=\"813,-215 813,-252 1037,-252 1037,-215 813,-215\"/>\n",
+       "<text text-anchor=\"start\" x=\"824\" y=\"-230.8\"
```
Comment (xpilasav) [inline retail/geopricing/main.ipynb:2424]: _outlets_feature_ not in italic fully
